chain/src/chain.rs

```diff
@@ -194,71 +190,67 @@ impl ChainService {
         self.process_block(block, Switch::NONE)
     }
 
+    fn make_fork_for_truncate(&self, target: &HeaderView, current_tip: &HeaderView) -> ForkChanges {
+        let mut fork = ForkChanges::default();
+        let store = self.shared.store();
+        for bn in (target.number() + 1)..=current_tip.number() {
+            let hash = store.get_block_hash(bn).expect("index checked");
+            let old_block = store.get_block(&hash).expect("index checked");
+            fork.detached_blocks.push_front(old_block);
+        }
+        fork
+    }
+
     // Truncate the main chain
     // Use for testing only
     pub(crate) fn truncate(&mut self, target_tip_hash: &Byte32) -> Result<(), Error> {
-        let snapshot = self.shared.snapshot();
+        let snapshot = Arc::clone(&self.shared.snapshot());
         assert!(snapshot.is_main_chain(target_tip_hash));
 
         let target_tip_header = snapshot.get_block_header(target_tip_hash).expect("checked");
-        let target_tip_number = target_tip_header.number();
+        let target_block_ext = snapshot.get_block_ext(target_tip_hash).expect("checked");
+        let target_epoch_ext = snapshot
+            .get_block_epoch_index(target_tip_hash)
+            .and_then(|index| snapshot.get_epoch_ext(&index))
+            .expect("checked");
+        let origin_proposals = snapshot.proposals();
+        let mut fork = self.make_fork_for_truncate(&target_tip_header, snapshot.tip_header());
 
-        // Detach blocks from main chain and remove from database
         let db_txn = self.shared.store().begin_transaction();
-        let mut cursor = snapshot.tip_hash();
-        while target_tip_hash != &cursor {
-            let block = snapshot.get_block(&cursor).expect("checked");
-            let block_number: Uint64 = block.number().pack();
-            db_txn.detach_block(&block)?;
-            detach_block_cell(&db_txn, &block)?;
-
-            // Remove the whole block data from database
-            let columns = vec![
-                COLUMN_INDEX,
-                COLUMN_BLOCK_HEADER,
-                COLUMN_BLOCK_BODY,
-                COLUMN_BLOCK_EXT,
-                COLUMN_BLOCK_PROPOSAL_IDS,
-                COLUMN_BLOCK_EPOCH,
-            ];
-            for column in columns {
-                db_txn.delete(column, cursor.as_slice())?;
```

Comment:
db_txn.delete(column, cursor.as_slice())?; for `COLUMN_BLOCK_BODY` is incorrect.

see https://github.com/nervosnetwork/ckb/pull/2120/files#diff-ed3597f2dbe252bad565f2898a1c8bd7R126-R141